File: src/main/java/org/lanternpowered/server/service/permission/base/LanternSubject.java
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.service.permission.base;

import org.spongepowered.api.service.context.Context;
import org.spongepowered.api.service.permission.MemorySubjectData;
import org.spongepowered.api.service.permission.PermissionService;
import org.spongepowered.api.service.permission.Subject;
import org.spongepowered.api.service.permission.SubjectData;
import org.spongepowered.api.service.permission.SubjectReference;
import org.spongepowered.api.util.Tristate;

import java.util.List;
import java.util.Optional;
import java.util.Set;

public abstract class LanternSubject implements Subject {

    public abstract PermissionService getService();

    @Override
    public MemorySubjectData getTransientSubjectData() {
        return getSubjectData();
    }

    @Override
    public abstract MemorySubjectData getSubjectData();

    @Override
    public boolean isSubjectDataPersisted() {
        return false;
    }

    @Override
    public SubjectReference asSubjectReference() {
        return getService().newSubjectReference(getContainingCollection().getIdentifier(), getIdentifier());
    }

    @Override
    public boolean hasPermission(Set<Context> contexts, String permission) {
        return getPermissionValue(contexts, permission) == Tristate.TRUE;
    }

    @Override
    public Tristate getPermissionValue(Set<Context> contexts, String permission) {
        return getDataPermissionValue(getTransientSubjectData(), permission);
    }

    protected Tristate getDataPermissionValue(MemorySubjectData subject, String permission) {
        Tristate res = subject.getNodeTree(SubjectData.GLOBAL_CONTEXT).get(permission);
        if (res == Tristate.UNDEFINED) {
            for (SubjectReference parent : subject.getParents(SubjectData.GLOBAL_CONTEXT)) {
                res = parent.resolve().join().getPermissionValue(SubjectData.GLOBAL_CONTEXT, permission);
                if (res != Tristate.UNDEFINED) {
                    return res;
                }
            }
        }
        return res;
    }

    @Override
    public boolean isChildOf(Set<Context> contexts, SubjectReference parent) {
        return getSubjectData().getParents(contexts).contains(parent);
    }

    @Override
    public List<SubjectReference> getParents(Set<Context> contexts) {
        return getSubjectData().getParents(contexts);
    }

    protected Optional<String> getDataOptionValue(MemorySubjectData subject, String option) {
        Optional<String> res = Optional.ofNullable(subject.getOptions(SubjectData.GLOBAL_CONTEXT).get(option));
        if (!res.isPresent()) {
            for (SubjectReference parent : subject.getParents(SubjectData.GLOBAL_CONTEXT)) {
                res = parent.resolve().join().getOption(SubjectData.GLOBAL_CONTEXT, option);
                if (res.isPresent()) {
                    return res;
                }
            }
        }
        return res;
    }

    @Override
    public Optional<String> getOption(Set<Context> contexts, String key) {
        return getDataOptionValue(getTransientSubjectData(), key);
    }

    @Override
    public Set<Context> getActiveContexts() {
        return SubjectData.GLOBAL_CONTEXT;
    }
}

File: src/main/kotlin/org/lanternpowered/server/service/permission/LanternSubject.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.service.permission

import org.spongepowered.api.service.context.Context
import org.spongepowered.api.service.permission.MemorySubjectData
import org.spongepowered.api.service.permission.PermissionService
import org.spongepowered.api.service.permission.Subject
import org.spongepowered.api.service.permission.SubjectData
import org.spongepowered.api.service.permission.SubjectReference
import org.spongepowered.api.util.Tristate
import java.util.Optional

abstract class LanternSubject : Subject {

    abstract val service: PermissionService

    override fun getTransientSubjectData(): MemorySubjectData = this.subjectData

    abstract override fun getSubjectData(): MemorySubjectData

    override fun isSubjectDataPersisted(): Boolean = false

    override fun asSubjectReference(): SubjectReference =
            this.service.newSubjectReference(this.containingCollection.identifier, this.identifier)

    override fun hasPermission(contexts: Set<Context>, permission: String): Boolean =
            this.getPermissionValue(contexts, permission) == Tristate.TRUE

    override fun getPermissionValue(contexts: Set<Context>, permission: String): Tristate =
            this.getDataPermissionValue(this.transientSubjectData, permission)

    protected fun getDataPermissionValue(subject: MemorySubjectData, permission: String?): Tristate {
        var res = subject.getNodeTree(SubjectData.GLOBAL_CONTEXT)[permission]
        if (res === Tristate.UNDEFINED) {
            for (parent in subject.getParents(SubjectData.GLOBAL_CONTEXT)) {
                res = parent.resolve().join().getPermissionValue(SubjectData.GLOBAL_CONTEXT, permission)
                if (res !== Tristate.UNDEFINED) {
                    return res
                }
            }
        }
        return res
    }

    override fun isChildOf(contexts: Set<Context>, parent: SubjectReference): Boolean =
            this.subjectData.getParents(contexts).contains(parent)

    override fun getParents(contexts: Set<Context>): List<SubjectReference> =
            this.subjectData.getParents(contexts)

    protected fun getDataOptionValue(subject: MemorySubjectData, option: String): Optional<String> {
        var result = Optional.ofNullable(subject.getOptions(SubjectData.GLOBAL_CONTEXT)[option])
        if (!result.isPresent) {
            for (parent in subject.getParents(SubjectData.GLOBAL_CONTEXT)) {
                result = parent.resolve().join().getOption(SubjectData.GLOBAL_CONTEXT, option)
                if (result.isPresent)
                    return result
            }
        }
        return result
    }

    override fun getOption(contexts: Set<Context>, key: String): Optional<String> = this.getDataOptionValue(transientSubjectData, key)

    override fun getActiveContexts(): Set<Context> = SubjectData.GLOBAL_CONTEXT
}
